### ISSUE TITLE
Optional properties on narrowed type missing

### ISSUE THREAD

**Issue by m-rutter** [gvergnaud/ts-pattern#20]
Came across what I think is a bug when it comes to the narrowed types with only additional optional properties than what you are matching on. A work around is you include the additional property as part of the match in some way as I did in the second branch.

Version 2.4.0, tsc 4.2.3

```ts
import {match, not} from 'ts-pattern';

type Foo = 
{type: "test", id?: string} |
{type: "test2", id?: string, otherProp: string} |
{type: "test3", id?: string, otherProp?: string}

declare const foo: Foo

match(foo)
.with({type: "test"}, ({id}) => {}) // Property 'id' does not exist on type '{ type: "test"; }'.(2339)
.with({type: "test", id: not(undefined)}, ({id}) => {}) // ok
.with({type: "test2"}, ({id}) => {}) // ok
.with({type: "test3"}, ({id}) => {}) // Property 'id' does not exist on type '{ type: "test"; }'.(2339)

match(foo)
.exhaustive()
.with({type: "test"}, ({id}) => {}) // Property 'id' does not exist on type '{ type: "test"; }'.(2339)
.with({type: "test", id: not(undefined)}, ({id}) => {}) // ok
.with({type: "test2"}, ({id}) => {}) // ok
.with({type: "test3"}, ({id}) => {}) // Property 'id' does not exist on type '{ type: "test"; }'.(2339)
```

[Playground link](https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=21&pc=37#code/JYWwDg9gTgLgBAbxAQxgYwBYBo4DsIwC+cAZlBCHAOQwDOAtGKjAKZS5UDcAUNzAJ5gWcAGIQIcALxxuCAUIBccAEStaMZTmAATAPxL1UYLgDmxAD6z5LJapbqATJrg79cQ8ZM4CGNgAVyMAMYI1MLK0EbFTUYAGZnV2DQrzgff0C3DzDebRY0ABtkKGE0CFx1UnElMQheFHQMAAoScQBKbgA6AHdgGCa5SNsY5UIcRoQdQlapAD5EKc6evvHrIfsNLW0lfBhGgFdcXJJjFm1W0bhxyenJOYQF7t7+1ej1pwur7SnZ+fbH5YGile6niHwmXxudwW3HqmGabU6LAAHhhkHt1MAAG4sRp-JbPQbAjRg64-e54p4rQl2dQJLZ4Aj7Q4sY64U7nMbg763X6LSmAqI0mDvTmknnkvkAl5C0GiiFkqZAA)

**Comment by m-rutter** [comment]
Ok, so this appears to be fixed in `3.0.2`.

**Comment by m-rutter** [comment]
I will leave this open just in case you weren't aware of it and it still exists in some form. You didn't mention any bug fixes related this.

**Comment by m-rutter** [comment]
I previously reported that this is fixed in `3.0.2`, but I was incorrect. I am able to reproduce in the latest major version.

**Comment by gvergnaud** [comment]
Thanks for reporting, I'll look into this soon

**Comment by gvergnaud** [comment]
Fixed in `v3.0.3`, [see this commit](https://github.com/gvergnaud/ts-pattern/commit/bee623d1c1a8d850053ae6fd1692c0416eb15a9c) for more details

**Comment by julienavert** [comment]
Hey there, sorry for digging up an old issue (tell me if I should create a new one), but I think I am experiencing this issue with latest version (`v4.1.3`)

The setup is slightly different: [typescriptlang playground](https://www.typescriptlang.org/play?target=99&module=1&pretty=true&allowSyntheticDefaultImports=true&ts=4.7.4#code/JYWwDg9gTgLgBAbziAhjAxgCzgXzgMyghDgHIYBnAWjDRgFMoA7UgKFfqYFcSBlY+gFFuJBKzgS4AMQgQ4AXjIyIpADTjJAIRRQFZbVDY52MAJ5h6cfiHoB5AEYAreuniKxkuCgD8ALisCwjwA3BoS9n5wFDBQwEwA5qGe6JH2sgA29ChMocas6BBM0XBxYFww-tZ2Ti5uiGFelYEiAHTK6p72-qRMEPCgkLDZMEbsBUXwUPQUXOl1qBiYABSl5QCUDS0A7sAwy0goTTZBIG2yuKpwAG4o6VyW8gB8cAAGyv4AJAg3d-Qt9jgXhtPNtdvtGgFjq0DBdrrd7gpni8DJ9vvC-gCgZsdnslgd-FwmAATej4OL0Imwn4Ip6vAA8hJJZKYFMeqOpGMBwMkLXoAA9MCguNFgFd6Es1qF8oUKBBMi10hB4kspjM5pLWEA)

inlined:

```ts
enum SomeEnum {
    Foo = 'Foo',
    Bar = 'Bar'
}

type SomeObject = {
    a?: SomeEnum;
    b?: string;
    c?: boolean;
}

const input: SomeObject = {
    a: SomeEnum.Foo,
    b: 'not important'
}

const result = match(input)
    .with({ a: SomeEnum.Foo }, value => `Foo: ${value.b}`)
    .with({ a: SomeEnum.Bar }, value => `Bar: ${value.b}`)
    .with({ a: undefined }, value => `<undefined>: ${value.b}`)
    .exhaustive();
```

In the playground you can see that the `with` narrowing is removing all optional properties from the input type

![image](https://user-images.githubusercontent.com/87069068/216660721-b7442916-6748-4f25-be3c-f8c7bdb8d35c.png)

Am I using the lib incorrectly ? I personally would expect the narrowed type for the first case to be

```ts
{
    a: SomeEnum.Foo;
    b?: string;
    c?: boolean;
}
```

or maybe even

```ts
{
    a: SomeEnum.Foo;
    b: string | undefined;
    c: boolean | undefined;
}
```

It is interesting to note that if the initial type for the property `b` is `string | undefined` instead of the property being optional, it works as expected (it even fixes the issue for **other** properties!) [typescriptlang playground](https://www.typescriptlang.org/play?target=99&module=1&pretty=true&allowSyntheticDefaultImports=true&ts=4.7.4#code/JYWwDg9gTgLgBAbziAhjAxgCzgXzgMyghDgHIYBnAWjDRgFMoA7UgKFfqYFcSBlY+gFFuJBKzgS4AMQgQ4AXjIyIpADTjJAIRRQFZbVDY52MAJ5h6cfiHoB5AEYAreuniKxkuCgD8ALisCwjwA3BoS9v4UMFDATADmcAA+cFxMACb0+LH0aaGe6H5w9rIANvQoTKHGrOgQTFFwsWBcMP7Wdk4ubohhXm2BIgB0yuqeEWRMEPCgkLAVMEbstfXwUPQUXCXdqBiYABRNLQCUvYMA7sAw+0go-TZBIMOyuKpwAG4oJVyW8gB8cAADZT+AAkCA+X3og3sOABJ0850u1z6AXuQwML3en2+Cn+AIMoPB2KhMLhpwuVz2N38qQyWSYOUxEJxf0BAB5aZlsmlfoTmSTYfDJIN6AAPTAoLhRYBveh7I6hGp1CgQMqDEoQOJ7NYbLYK1hAA)

inlined:

```ts
enum SomeEnum {
    Foo = 'Foo',
    Bar = 'Bar'
}

type SomeObject = {
    a?: SomeEnum;
    b: string | undefined;
    c?: boolean;
}

const input: SomeObject = {
    a: SomeEnum.Foo,
    b: 'not important'
}

const result = match(input)
    .with({ a: SomeEnum.Foo }, value => `Foo: ${value.b}`)
    .with({ a: SomeEnum.Bar }, value => `Bar: ${value.b}`)
    .with({ a: undefined }, value => `<undefined>: ${value.b}`)
    .exhaustive();
```

![image](https://user-images.githubusercontent.com/87069068/216661819-4a9292d4-6145-4e7a-802f-7a6b3bc9b4fa.png)
